src/profile-logic/import/instruments/index.js

```diff
@@ -0,0 +1,910 @@
+      const timestamp = bulkstore.readUint48();
+      if (timestamp === 0) {
+        break;
+      }
+
+      const threadID = bulkstore.readUint32();
+
+      bulkstore.skip(bytesPerEntry - 6 - 4 - 4);
+      const backtraceID = bulkstore.readUint32();
+      samples.push({ timestamp, threadID, backtraceID });
+    }
+    return samples;
+  }
+  throw new Error('Could not find sample list');
+}
+
+// This function extracts core directory file which contains information about samples
+// path: tree.subdirectories.corespace.subdirectories.run${runNumber}.core
+function getCoreDirForRun(
+  tree: TraceDirectoryTree,
+  selectedRun: number
+): TraceDirectoryTree {
+  const corespace = getOrThrow(tree.subdirectories, 'corespace');
+  const corespaceRunDir = getOrThrow(
+    corespace.subdirectories,
+    `run${selectedRun}`
+  );
+  return getOrThrow(corespaceRunDir.subdirectories, 'core');
+}
+
+// This function mainly extracts two entities: samples and arrays
+// Here arrays contains all the information about stack trace at a given timestamp.
+// Each samples has a field named 'backtraceID' which is an index into arrays
+// Iterating recursively into arrays by given backtraceID it extracts a backtraceStack for each sample
+async function getSamples(args: {
+  tree: TraceDirectoryTree,
+  addressToFrameMap: Map<number, InstrumentsFrameInfo>,
+  runNumber: number,
+}): Promise<Array<InstrumentsSample>> {
+  const { tree, addressToFrameMap, runNumber } = args;
+  const core = getCoreDirForRun(tree, runNumber);
+  const samples = await getRawSampleList(core);
+  const arrays = await getIntegerArrays(samples, core);
+  const backtraceIDtoStack = new Map<number, number[]>();
+
+  function appendRecursive(k: number, stack: Array<number>) {
+    const frame = addressToFrameMap.get(k);
+    if (frame) {
+      stack.push(k);
+      return;
+    }
+
+    const arrayIndex = k & 0xffffffff;
+    if (arrayIndex in arrays) {
+      for (const addr of arrays[arrayIndex]) {
+        appendRecursive(addr, stack);
+      }
+      return;
+    }
+
+    // Fallback: We are not able to find the address, so we will just
+    // display the address instead of frame information
+    const rawAddressFrame: InstrumentsFrameInfo = {
+      key: k,
+      name: `0x${zeroPad(k.toString(16), 16)}`,
+    };
+    addressToFrameMap.set(k, rawAddressFrame);
+    stack.push(k);
+  }
+
+  for (const sample of samples) {
+    getOrInsert(backtraceIDtoStack, sample.backtraceID, id => {
+      const stack: Array<number> = [];
+      appendRecursive(id, stack);
+      stack.reverse();
+      return stack;
+    });
+  }
+
+  for (const sample of samples) {
+    sample.backtraceStack = backtraceIDtoStack.get(sample.backtraceID);
+  }
+
+  return Promise.resolve(samples);
+}
+
+// This function reads the 'form.template' file which contains all the important information about
+// addressToFrameMap and Instruments' version
+async function readFormTemplateFile(tree) {
```

Comment:
please add flow types here

Reply:
Added the type. Resolving the comment. :) @julienw